### PR TITLE
Add gap junction and cell documentation

### PR DESCRIPTION
**Pull request by noraabiakar**
Updates the docs:
* Replaces Arbor Model::Common Types with Arbor Model::Concepts, that introduces the concept of cells, connections and gap junctions. 
* Adds gap junction documentation to cpp:recipe
* Documents the implicit rule forcing cells connected by gap junctions to be in the same cell group.

### CODE REVIEW

File: doc/cpp_recipe.rst [hunk 1]
```diff
@@ -154,3 +170,19 @@ Class Documentation
 
         Delay of the connection (milliseconds).
 
+.. cpp:class:: gap_junction_connections
+
+    Describes a gap junction between two gap junction sites on two cells.
+    Gap junction sites are represented by cell_member_type.
+
+    .. cpp:member:: cell_member_type local
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply doc/cpp_recipe.rst]: That's what Sam and I settled on back when we were working on the gap junctions PR. I agree that it's a bit confusing. I can change the naming in a separate PR, do you have any suggestions?

[the remark this answers is not in the document]
Comment (halfflat) [reply doc/cpp_recipe.rst]: We could make it `cell_member_type peer[2]`.

[the remark this answers is not in the document]
Comment (noraabiakar) [reply doc/cpp_recipe.rst]: ```cell_member_type peer[2]``` works for me. I'll fix the docs and put up another PR to change the source code.

File: doc/model_concepts.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                                                      * **lif**: leaky-integrate and fire neuron.
+                                                      * **spiking**: spike source from values inserted via
+                                                        description.
+                                                      * **benchmark**: proxy cell used for benchmarking.
+    ========================  ======================  ===========================================================
+
+1. **Cable**
+
+   Cable cells are morphologically detailed cells represented as branching linear 1D segments. They allow the addition of
+   point and density mechanisms. They allow having *gap junctions* and *connections* to other cable cells.
```
Comment (akuesters) [inline doc/model_concepts.rst]: Does it make sense to include `:cpp:` references in the generic description? I would rather link to the C++ (later Python) specific documentation and keep it generic here.

Comment (bcumming) [reply doc/model_concepts.rst]: Good point. We have to find a way to define the "generic" type, e.g. cell_member_type, and refer to it here.

File: doc/cpp_recipe.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (bcumming) [reply doc/cpp_recipe.rst]: That is better. It is a "symmetric" relationship, but I am struggling to think fo better words to use (left-right) (first-second) (endpoint1, endpoint2)... all the ones I can think of a pretty terrible.

File: doc/model_concepts.rst (hunk 2 above)
Comment (bcumming) [reply doc/model_concepts.rst]: Nora found a nice way to define a new "generic" reference type to use for definitions of types in the model documentation.

**Pull request merged**